README.md
# LEARNING C/C++ BY DEVELOPING A POKEMON GAME 

Currently developing a pokemon game in C and C++
Adding new changes every week !
Things done till now :
1. Build standard Map with intersecting paths, pokemon centers and pokemarts which are all randomly generated with multiple kinds of terrains for the players to travel through
2. Connected 401 * 401 maps to make a world with different maps for player to traverse through using keyboards inputs, player can also fly to a map in a different region
3. Added a playing character to the map and generated shortest distances for all "hiker" and "rival" (non-playing characters or NPC's) till the playing characters based on different terrains having differing distance values for various kinds of NPC's.
4. Developing movement constraints for all the non-player characters and running a simulation until user asks to exit. Next steps include adding UI, making player move through user inputs and converting parts of code to C++!



Changelog.txt
commit 106b8ecfa069e7fb85f8a18e3ad488eb64f5a398
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 22 20:02:58 2023 -0500

     Adding new comments and going through the code and checking all implemented methods in the class once more

commit 1506e2e57ed1b72e8621a347b8310db8e6c048e6
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 22 19:37:10 2023 -0500

     Changing make file and checking if code works

commit fded1deb33d661643a431ece10efad9ffe4f5990
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 22 19:08:39 2023 -0500

     Restructuring folders to check output of only what we need

commit 6eaba2bd8a6f2fbf8c9fe821aeb3eec4aa512a70
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 22 19:04:59 2023 -0500

     Fixed errors for player movement and added battle commands

commit 4611177427aeeda48b63ba2a38d8d90e3e337c5f
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 22 18:07:03 2023 -0500

     Fixing PC movement and hiker and rival maps should change according to player

commit 3dce710e9dd109ff84ef32a76a77233434ebb2f0
Merge: b4bd356 2fcbaaa
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 22 17:03:15 2023 -0500

    Merge branch 'main' of https://github.com/akhilnev/327

commit b4bd356bc00bff79b8b76464cb375a514d948b84
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 22 17:03:07 2023 -0500

     Adding trial 1.05 and trying to Switch to Professor's 1.04 Solution code as it looks more modular and efficient at this point

commit 2fcbaaac6aaa406f7eb8cd0c7190a4db09c63fbf
Author: Akhilesh Nevatia <106297042+akhilnev@users.noreply.github.com>
Date:   Sat Oct 14 02:17:18 2023 -0500

    Update README.md

commit b216176cf006ef45749e3307b2fa858052c3da50
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Mon Oct 9 01:39:04 2023 -0500

     Added working files

commit 5fcac7d4e3cc83de35897a1d49605df4bf6ae275
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 8 23:26:18 2023 -0500

     Still figuring out h and r path to @

commit d7b4d5b46c05444aeaac9a18acfe75608cd6ed42
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 8 21:14:12 2023 -0500

     Still validating routes for hiker and rival

commit c35020efddcc3628b20492328529f5a3ccb3fe71
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 8 20:15:35 2023 -0500

     Cecking hiker and rival movement conditions

commit 4c99646eedd34008aeb76994a5f923702c9ee120
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 8 17:47:24 2023 -0500

     Added logic for all 6 NPC's and am trying to fix conflicting PQ errors between Dijkstra PQ and Movement PQ

commit 5232bfc053f0530588ccf9ab0454d3a4f0663c10
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 8 15:46:55 2023 -0500

     Commiting readme , and a.out ,and other files made

commit e7959983faf6db70a431b971ab3ff64e1ee08d67
Merge: acd2a2b e022e70
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 8 15:45:46 2023 -0500

    Merge branch 'main' of https://github.com/akhilnev/327

commit acd2a2bd8482c636a142adca3afd154b8eb6936e
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 8 15:41:28 2023 -0500

     Adding the NPC'S and also adding movement routes

commit e022e70c21f00d60f6bcc0adf1695ca4cdf74585
Author: Akhilesh Nevatia <106297042+akhilnev@users.noreply.github.com>
Date:   Sun Oct 1 11:39:09 2023 -0500

    Update README.md

commit 74b072bd6fdbf38e793f150f6ff66f819f719fcb
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Oct 1 01:37:06 2023 -0500

     Adding Rivals map , makefile and Working on Readme

commit 82b382d5a029c45999426d411d15bb59d2f0391f
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sat Sep 30 20:48:27 2023 -0500

     Adding new implementation of map

commit a7d074b51d0bdc58c535e30ea10575e6727313fd
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Sep 24 00:05:14 2023 -0500

     Adding Dijkstra function and trying to debug it at the same time

commit ef89eb3c1e350f8a774640bdeaa139f3930f5220
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sat Sep 23 20:26:53 2023 -0500

     Adding Dijkstra implementation , function need to fix infinite loop

commit 006294f729c41886b737f62510255784985997c1
Merge: fb41f20 f75b3cb
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sat Sep 23 18:16:33 2023 -0500

    Merge branch 'main' of https://github.com/akhilnev/327

commit fb41f200ea52bf8abc3de2f467a428116dd963b6
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sat Sep 23 18:16:20 2023 -0500

     Starting project 1.03 by initializing PC and randomly assigning position in path, adding hiker and rival dijkstra maps which I will now implement

commit f75b3cb7533992ff95560c2fb11f9043e1d3edae
Author: Akhilesh Nevatia <106297042+akhilnev@users.noreply.github.com>
Date:   Thu Sep 21 12:30:09 2023 -0500

    Create README.md

commit 7190a2538f8337bdf6c579dc782fba9d003920e6
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Sep 17 22:16:09 2023 -0500

     Adding current map coordinate printer

commit 044dc1fca9d924aaaaae45a8d690fff84c857d29
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Sep 17 22:05:19 2023 -0500

    Adding fixed changes

commit 66f1b41e02875f67407410d2568217266171d12b
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>
Date:   Sun Sep 17 21:27:58 2023 -0500

     Adding Readme , make file and testing boundry conditions

commit 74b35feab1b12dd55a66697dc1e80af1a11eb849
Author: akhilesh.nevatia@gmail.com <akhilesh.nevatia@gmail.com>

Nevatia_Akhilesh-Assignment1.06/io.cpp
#include <unistd.h>
#include <ncurses.h>
#include <ctype.h>
#include <stdlib.h>
#include <limits.h>

#include "io.h"
#include "character.h"
#include "poke327.h"

typedef struct io_message {
  /* Will print " --more-- " at end of line when another message follows. *
   * Leave 10 extra spaces for that.                                      */
  char msg[71];
  struct io_message *next;
} io_message_t;

static io_message_t *io_head, *io_tail;

void io_init_terminal(void)
{
  initscr();
  raw();
  noecho();
  curs_set(0);
  keypad(stdscr, TRUE);
  start_color();
  init_pair(COLOR_RED, COLOR_RED, COLOR_BLACK);
  init_pair(COLOR_GREEN, COLOR_GREEN, COLOR_BLACK);
  init_pair(COLOR_YELLOW, COLOR_YELLOW, COLOR_BLACK);
  init_pair(COLOR_BLUE, COLOR_BLUE, COLOR_BLACK);
  init_pair(COLOR_MAGENTA, COLOR_MAGENTA, COLOR_BLACK);
  init_pair(COLOR_CYAN, COLOR_CYAN, COLOR_BLACK);
  init_pair(COLOR_WHITE, COLOR_WHITE, COLOR_BLACK);
}

void io_reset_terminal(void)
{
  endwin();

  while (io_head) {
    io_tail = io_head;
    io_head = io_head->next;
    free(io_tail);
  }
  io_tail = NULL;
}

void io_queue_message(const char *format, ...)
{
  io_message_t *tmp;
  va_list ap;

  if (!(tmp = (io_message_t *) malloc(sizeof (*tmp)))) {
    perror("malloc");
    exit(1);
  }

  tmp->next = NULL;

  va_start(ap, format);

  vsnprintf(tmp->msg, sizeof (tmp->msg), format, ap);

  va_end(ap);

  if (!io_head) {
    io_head = io_tail = tmp;
  } else {
    io_tail->next = tmp;
    io_tail = tmp;
  }
}

static void io_print_message_queue(uint32_t y, uint32_t x)
{
  while (io_head) {
    io_tail = io_head;
    attron(COLOR_PAIR(COLOR_CYAN));
    mvprintw(y, x, "%-80s", io_head->msg);
    attroff(COLOR_PAIR(COLOR_CYAN));
    io_head = io_head->next;
    if (io_head) {
      attron(COLOR_PAIR(COLOR_CYAN));
      mvprintw(y, x + 70, "%10s", " --more-- ");
      attroff(COLOR_PAIR(COLOR_CYAN));
      refresh();
      getch();
    }
    free(io_tail);
  }
  io_tail = NULL;
}

/**************************************************************************
 * Compares trainer distances from the PC according to the rival distance *
 * map.  This gives the approximate distance that the PC must travel to   *
 * get to the trainer (doesn't account for crossing buildings).  This is  *
 * not the distance from the NPC to the PC unless the NPC is a rival.     *
 *                                                                        *
 * Not a bug.                                                             *
 **************************************************************************/
static int compare_trainer_distance(const void *v1, const void *v2)
{
  const character *const *c1 = (const character * const *) v1;
  const character *const *c2 = (const character * const *) v2;

  return (world.rival_dist[(*c1)->pos[dim_y]][(*c1)->pos[dim_x]] -
          world.rival_dist[(*c2)->pos[dim_y]][(*c2)->pos[dim_x]]);
}

static character *io_nearest_visible_trainer()
{
  character **c, *n;
  uint32_t x, y, count;

  c = (character **) malloc(world.cur_map->num_trainers * sizeof (*c));

  /* Get a linear list of trainers */
  for (count = 0, y = 1; y < MAP_Y - 1; y++) {
    for (x = 1; x < MAP_X - 1; x++) {
      if (world.cur_map->cmap[y][x] && world.cur_map->cmap[y][x] !=
          &world.pc) {
        c[count++] = world.cur_map->cmap[y][x];
      }
    }
  }

  /* Sort it by distance from PC */
  qsort(c, count, sizeof (*c), compare_trainer_distance);

  n = c[0];

  free(c);

  return n;
}

void io_display()
{
  uint32_t y, x;
  character *c;

  clear();
  for (y = 0; y < MAP_Y; y++) {
    for (x = 0; x < MAP_X; x++) {
      if (world.cur_map->cmap[y][x]) {
        mvaddch(y + 1, x, world.cur_map->cmap[y][x]->symbol);
      } else {
        switch (world.cur_map->map[y][x]) {
        case ter_boulder:
          attron(COLOR_PAIR(COLOR_MAGENTA));
          mvaddch(y + 1, x, BOULDER_SYMBOL);
          attroff(COLOR_PAIR(COLOR_MAGENTA));
          break;
        case ter_mountain:
          attron(COLOR_PAIR(COLOR_MAGENTA));
          mvaddch(y + 1, x, MOUNTAIN_SYMBOL);
          attroff(COLOR_PAIR(COLOR_MAGENTA));
          break;
        case ter_tree:
          attron(COLOR_PAIR(COLOR_GREEN));
          mvaddch(y + 1, x, TREE_SYMBOL);
          attroff(COLOR_PAIR(COLOR_GREEN));
          break;
        case ter_forest:
          attron(COLOR_PAIR(COLOR_GREEN));
          mvaddch(y + 1, x, FOREST_SYMBOL);
          attroff(COLOR_PAIR(COLOR_GREEN));
          break;
        case ter_path:
          attron(COLOR_PAIR(COLOR_YELLOW));
          mvaddch(y + 1, x, PATH_SYMBOL);
          attroff(COLOR_PAIR(COLOR_YELLOW));
          break;
        case ter_gate:
          attron(COLOR_PAIR(COLOR_YELLOW));
          mvaddch(y + 1, x, GATE_SYMBOL);
          attroff(COLOR_PAIR(COLOR_YELLOW));
          break;
        case ter_mart:
          attron(COLOR_PAIR(COLOR_BLUE));
          mvaddch(y + 1, x, POKEMART_SYMBOL);
          attroff(COLOR_PAIR(COLOR_BLUE));
          break;
        case ter_center:
          attron(COLOR_PAIR(COLOR_RED));
          mvaddch(y + 1, x, POKEMON_CENTER_SYMBOL);
          attroff(COLOR_PAIR(COLOR_RED));
          break;
        case ter_grass:
          attron(COLOR_PAIR(COLOR_GREEN));
          mvaddch(y + 1, x, TALL_GRASS_SYMBOL);
          attroff(COLOR_PAIR(COLOR_GREEN));
          break;
        case ter_clearing:
          attron(COLOR_PAIR(COLOR_GREEN));
          mvaddch(y + 1, x, SHORT_GRASS_SYMBOL);
          attroff(COLOR_PAIR(COLOR_GREEN));
          break;
        case ter_water:
          attron(COLOR_PAIR(COLOR_CYAN));
          mvaddch(y + 1, x, WATER_SYMBOL);
          attroff(COLOR_PAIR(COLOR_CYAN));
          break;
        default:
          attron(COLOR_PAIR(COLOR_CYAN));
          mvaddch(y + 1, x, ERROR_SYMBOL);
          attroff(COLOR_PAIR(COLOR_CYAN)); 
       }
      }
    }
  }

  mvprintw(23, 1, "PC position is (%2d,%2d) on map %d%cx%d%c.",
           world.pc.pos[dim_x],
           world.pc.pos[dim_y],
           abs(world.cur_idx[dim_x] - (WORLD_SIZE / 2)),
           world.cur_idx[dim_x] - (WORLD_SIZE / 2) >= 0 ? 'E' : 'W',
           abs(world.cur_idx[dim_y] - (WORLD_SIZE / 2)),
           world.cur_idx[dim_y] - (WORLD_SIZE / 2) <= 0 ? 'N' : 'S');
  mvprintw(22, 1, "%d known %s.", world.cur_map->num_trainers,
           world.cur_map->num_trainers > 1 ? "trainers" : "trainer");
  mvprintw(22, 30, "Nearest visible trainer: ");
  if ((c = io_nearest_visible_trainer())) {
    attron(COLOR_PAIR(COLOR_RED));
    mvprintw(22, 55, "%c at vector %d%cx%d%c.",
             c->symbol,
             abs(c->pos[dim_y] - world.pc.pos[dim_y]),
             ((c->pos[dim_y] - world.pc.pos[dim_y]) <= 0 ?
              'N' : 'S'),
             abs(c->pos[dim_x] - world.pc.pos[dim_x]),
             ((c->pos[dim_x] - world.pc.pos[dim_x]) <= 0 ?
              'W' : 'E'));
    attroff(COLOR_PAIR(COLOR_RED));
  } else {
    attron(COLOR_PAIR(COLOR_BLUE));
    mvprintw(22, 55, "NONE.");
    attroff(COLOR_PAIR(COLOR_BLUE));
  }

  io_print_message_queue(0, 0);

  refresh();
}

uint32_t io_teleport_pc(pair_t dest)
{
  /* Just for fun. And debugging.  Mostly debugging. */

  do {
    dest[dim_x] = rand_range(1, MAP_X - 2);
    dest[dim_y] = rand_range(1, MAP_Y - 2);
  } while (world.cur_map->cmap[dest[dim_y]][dest[dim_x]]                  ||
           move_cost[char_pc][world.cur_map->map[dest[dim_y]]
                                                [dest[dim_x]]] == INT_MAX ||
           world.rival_dist[dest[dim_y]][dest[dim_x]] < 0);

  return 0;
}

static void io_scroll_trainer_list(char (*s)[40], uint32_t count)
{
  uint32_t offset;
  uint32_t i;

  offset = 0;

  while (1) {
    for (i = 0; i < 13; i++) {
      mvprintw(i + 6, 19, " %-40s ", s[i + offset]);
    }
    switch (getch()) {
    case KEY_UP:
      if (offset) {
        offset--;
      }
      break;
    case KEY_DOWN:
      if (offset < (count - 13)) {
        offset++;
      }
      break;
    case 27:
      return;
    }

  }
}

static void io_list_trainers_display(npc **c,
                                     uint32_t count)
{
  uint32_t i;
  char (*s)[40]; /* pointer to array of 40 char */

  s = (char (*)[40]) malloc(count * sizeof (*s));

  mvprintw(3, 19, " %-40s ", "");
  /* Borrow the first element of our array for this string: */
  snprintf(s[0], 40, "You know of %d trainers:", count);
  mvprintw(4, 19, " %-40s ", *s);
  mvprintw(5, 19, " %-40s ", "");

  for (i = 0; i < count; i++) {
    snprintf(s[i], 40, "%16s %c: %2d %s by %2d %s",
             char_type_name[c[i]->ctype],
             c[i]->symbol,
             abs(c[i]->pos[dim_y] - world.pc.pos[dim_y]),
             ((c[i]->pos[dim_y] - world.pc.pos[dim_y]) <= 0 ?
              "North" : "South"),
             abs(c[i]->pos[dim_x] - world.pc.pos[dim_x]),
             ((c[i]->pos[dim_x] - world.pc.pos[dim_x]) <= 0 ?
              "West" : "East"));
    if (count <= 13) {
      /* Handle the non-scrolling case right here. *
       * Scrolling in another function.            */
      mvprintw(i + 6, 19, " %-40s ", s[i]);
    }
  }

  if (count <= 13) {
    mvprintw(count + 6, 19, " %-40s ", "");
    mvprintw(count + 7, 19, " %-40s ", "Hit escape to continue.");
    while (getch() != 27 /* escape */)
      ;
  } else {
    mvprintw(19, 19, " %-40s ", "");
    mvprintw(20, 19, " %-40s ",
             "Arrows to scroll, escape to continue.");
    io_scroll_trainer_list(s, count);
  }

  free(s);
}

static void io_list_trainers()
{
  npc **c;
  uint32_t x, y, count;

  c = (npc **) malloc(world.cur_map->num_trainers * sizeof (*c));

  /* Get a linear list of trainers */
  for (count = 0, y = 1; y < MAP_Y - 1; y++) {
    for (x = 1; x < MAP_X - 1; x++) {
      if (world.cur_map->cmap[y][x] && world.cur_map->cmap[y][x] !=
          &world.pc) {
        c[count++] = dynamic_cast<npc *> (world.cur_map->cmap[y][x]);
      }
    }
  }

  /* Sort it by distance from PC */
  qsort(c, count, sizeof (*c), compare_trainer_distance);

  /* Display it */
  io_list_trainers_display(c, count);
  free(c);

  /* And redraw the map */
  io_display();
}

void io_pokemart()
{
  mvprintw(0, 0, "Welcome to the Pokemart.  Could I interest you in some Pokeballs?");
  refresh();
  getch();
}

void io_pokemon_center()
{
  mvprintw(0, 0, "Welcome to the Pokemon Center.  How can Nurse Joy assist you?");
  refresh();
  getch();
}

void io_battle(character *aggressor, character *defender)
{
  npc *n = (npc *) ((aggressor == &world.pc) ? defender : aggressor);

  io_display();
  mvprintw(0, 0, "Aww, how'd you get so strong?  You and your pokemon must share a special bond!");
  refresh();
  getch();

  n->defeated = 1;
  if (n->ctype == char_hiker || n->ctype == char_rival) {
    n->mtype = move_wander;
  }
}

uint32_t move_pc_dir(uint32_t input, pair_t dest)
{
  dest[dim_y] = world.pc.pos[dim_y];
  dest[dim_x] = world.pc.pos[dim_x];

  switch (input) {
  case 1:
  case 2:
  case 3:
    dest[dim_y]++;
    break;
  case 4:
  case 5:
  case 6:
    break;
  case 7:
  case 8:
  case 9:
    dest[dim_y]--;
    break;
  }
  switch (input) {
  case 1:
  case 4:
  case 7:
    dest[dim_x]--;
    break;
  case 2:
  case 5:
  case 8:
    break;
  case 3:
  case 6:
  case 9:
    dest[dim_x]++;
    break;
  case '>':
    if (world.cur_map->map[world.pc.pos[dim_y]][world.pc.pos[dim_x]] ==
        ter_mart) {
      io_pokemart();
    }
    if (world.cur_map->map[world.pc.pos[dim_y]][world.pc.pos[dim_x]] ==
        ter_center) {
      io_pokemon_center();
    }
    break;
  }

  if (world.cur_map->cmap[dest[dim_y]][dest[dim_x]]) {
    if (dynamic_cast<npc *> (world.cur_map->cmap[dest[dim_y]][dest[dim_x]]) &&
        ((npc *) world.cur_map->cmap[dest[dim_y]][dest[dim_x]])->defeated) {
      // Some kind of greeting here would be nice
      return 1;
    } else if ((dynamic_cast<npc *>
                (world.cur_map->cmap[dest[dim_y]][dest[dim_x]]))) {
      io_battle(&world.pc, world.cur_map->cmap[dest[dim_y]][dest[dim_x]]);
      // Not actually moving, so set dest back to PC position
      dest[dim_x] = world.pc.pos[dim_x];
      dest[dim_y] = world.pc.pos[dim_y];
    }
  }
  
  if (move_cost[char_pc][world.cur_map->map[dest[dim_y]][dest[dim_x]]] ==
      INT_MAX) {
    return 1;
  }

  return 0;
}

void io_teleport_world(pair_t dest)
{

int x = INT_MAX, y = INT_MAX;
char formatX[] = "%d";
char formatY[] = "%d";
  
world.cur_map->cmap[world.pc.pos[dim_y]][world.pc.pos[dim_x]] = NULL;

echo();
curs_set(1);

do {
    mvprintw(0, 0, "x b/w [-200,200]:");
    refresh();
    mvscanw(0, 21, formatX, &x);
} while (x < -200 || x > 200);

do {
    mvprintw(0, 0, "y b/w [-200,200]");
    refresh();
    mvscanw(0, 21, formatY, &y);
} while (y < -200 || y > 200);


  refresh();
  noecho();
  curs_set(0);

  x += 200;
  y += 200;

  world.cur_idx[dim_x] = x;
  world.cur_idx[dim_y] = y;

  new_map(1);
  io_teleport_pc(dest);
}

void io_handle_input(pair_t dest)
{
  uint32_t turn_not_consumed;
  int key;

  do {
    switch (key = getch()) {
    case '7':
    case 'y':
    case KEY_HOME:
      turn_not_consumed = move_pc_dir(7, dest);
      break;
    case '8':
    case 'k':
    case KEY_UP:
      turn_not_consumed = move_pc_dir(8, dest);
      break;
    case '9':
    case 'u':
    case KEY_PPAGE:
      turn_not_consumed = move_pc_dir(9, dest);
      break;
    case '6':
    case 'l':
    case KEY_RIGHT:
      turn_not_consumed = move_pc_dir(6, dest);
      break;
    case '3':
    case 'n':
    case KEY_NPAGE:
      turn_not_consumed = move_pc_dir(3, dest);
      break;
    case '2':
    case 'j':
    case KEY_DOWN:
      turn_not_consumed = move_pc_dir(2, dest);
      break;
    case '1':
    case 'b':
    case KEY_END:
      turn_not_consumed = move_pc_dir(1, dest);
      break;
    case '4':
    case 'h':
    case KEY_LEFT:
      turn_not_consumed = move_pc_dir(4, dest);
      break;
    case '5':
    case ' ':
    case '.':
    case KEY_B2:
      dest[dim_y] = world.pc.pos[dim_y];
      dest[dim_x] = world.pc.pos[dim_x];
      turn_not_consumed = 0;
      break;
    case '>':
      turn_not_consumed = move_pc_dir('>', dest);
      break;
    case 'Q':
      dest[dim_y] = world.pc.pos[dim_y];
      dest[dim_x] = world.pc.pos[dim_x];
      world.quit = 1;
      turn_not_consumed = 0;
      break;
      break;
    case 't':
      io_list_trainers();
      turn_not_consumed = 1;
      break;
    case 'p':
      /* Teleport the PC to a random place in the map.              */
      io_teleport_pc(dest);
      turn_not_consumed = 0;
      break;
    case 'f':
      /* Fly to any map in the world.                                */
      io_teleport_world(dest);
      turn_not_consumed = 0;
      break;    
    case 'q':
      /* Demonstrate use of the message queue.  You can use this for *
       * printf()-style debugging (though gdb is probably a better   *
       * option.  Not that it matters, but using this command will   *
       * waste a turn.  Set turn_not_consumed to 1 and you should be *
       * able to figure out why I did it that way.                   */
      io_queue_message("This is the first message.");
      io_queue_message("Since there are multiple messages, "
                       "you will see \"more\" prompts.");
      io_queue_message("You can use any key to advance through messages.");
      io_queue_message("Normal gameplay will not resume until the queue "
                       "is empty.");
      io_queue_message("Long lines will be truncated, not wrapped.");
      io_queue_message("io_queue_message() is variadic and handles "
                       "all printf() conversion specifiers.");
      io_queue_message("Did you see %s?", "what I did there");
      io_queue_message("When the last message is displayed, there will "
                       "be no \"more\" prompt.");
      io_queue_message("Have fun!  And happy printing!");
      io_queue_message("Oh!  And use 'Q' to quit!");

      dest[dim_y] = world.pc.pos[dim_y];
      dest[dim_x] = world.pc.pos[dim_x];
      turn_not_consumed = 0;
      break;
    default:
      /* Also not in the spec.  It's not always easy to figure out what *
       * key code corresponds with a given keystroke.  Print out any    *
       * unhandled key here.  Not only does it give a visual error      *
       * indicator, but it also gives an integer value that can be used *
       * for that key in this (or other) switch statements.  Printed in *
       * octal, with the leading zero, because ncurses.h lists codes in *
       * octal, thus allowing us to do reverse lookups.  If a key has a *
       * name defined in the header, you can use the name here, else    *
       * you can directly use the octal value.                          */
      mvprintw(0, 0, "Unbound key: %#o ", key);
      turn_not_consumed = 1;
    }
    refresh();
  } while (turn_not_consumed);
}
